Alex/use qc of uncertified block optimization

AlexHentschel opened the pull request:
Implements part 2 of https://github.com/onflow/flow-go/issues/4154
> 2. Forks can ingest `Proposals` as well as `CertifiedBlock`s. The finality logic in Forks accounts for the certifying QC.

## Approach taken in this PR

* I added a revised implementation of Forks, which I called `Forks2`
   * `Forks2` is not yet used by any business logic (will be done in subsequent PR)
   * extended tests to cover logic for adding `CertifiedBlock`s
* the old `Forks` and its unit tests remain in the code base, as they are still used
  * to avoid conflicts, I have moved the tests into the testing package `forks_test`
  * both `Forks` and its unit tests will be removed in subsequent PR

AlexHentschel commented on model/flow/block.go:
```diff
@@ -78,8 +78,8 @@ func (s BlockStatus) String() string {
-	Block *Block
-	QC    *QuorumCertificate
+	Block        *Block
+	CertifyingQC *QuorumCertificate
```
renamed this to `CertifyingQC` to avoid any confusion between the QC embedded in `Block` (i.e. `CertifiedBlock.Block.QC`, which is the QC for the _parent_) vs the `CertifyingQC` (which is the QC for the `CertifiedBlock` itself)

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+
+	// finalityProof holds the latest finalized block including the certified child as proof of finality.
+	// CAUTION: is nil, when Forks has not yet finalized any blocks beyond the finalized root block it was initialized with
+	finalityProof *FinalityProof //
```
```suggestion
	finalityProof *FinalityProof
```

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+	forest   forest.LevelledForest
+
+	trustedRoot          *model.CertifiedBlock
+	newestView           uint64 // newestView is the highest view of block proposal stored in Forks
```
do we still need `newestView`?

AlexHentschel replied:
💡 that is a good point:
* We currently call the `NewestView()` in the EventHandler [here](https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/eventhandler/event_handler.go#L322)
* Though, you are right that this logic will be removed, since we don't depend on iterating over forks to re-populate the PaceMaker. 

removed in https://github.com/onflow/flow-go/pull/4190/commits/4c70d562ea01013fbbf1cbf079e6a38ed6fc04f7

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+//     breaking the safety guarantees of HotStuff (or there is a critical bug / data
+//     corruption). Forks cannot recover from this exception.
+//   - All other errors are potential symptoms of bugs or state corruption.
+func (f *Forks2) UnverifiedAddCertifiedBlock(certifiedBlock *model.CertifiedBlock) error {
```
Why this function is public?

AlexHentschel replied:
🤷 No strong reason. We have similar exported methods in the old `Forks` implementation https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks/forks.go#L109 https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks/forks.go#L155 Though, we only expose `AddProposal` in the [`Forks` interface](https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks.go#L43).

When I originally implemented `Forks`, I also exposed the unverified methods to allow higher-level logic to choose which level of safety they desire. Conceptually, we don't need the additional validity check in `Forks`, because the consistency requirements should (in a perfect world 😅) be guaranteed by the compliance layer. 
But from a practical perspective:
* the additional validity check in `Forks` is probably really cheap
* the additional validity check is probably a very sensible sanity check (hardening the implementation)
* we anyway only expose the safe methods in the interface
* we don't use the `Unverified` methods at all (except within forks)

Based on these reasons, I totally agree with you that it would make the most sense to not export these methods.

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+	if err != nil {
+		return fmt.Errorf("storing block %v in Forks failed %w", certifiedBlock.Block.BlockID, err)
+	}
+	err = f.checkForConflictingQCs(certifiedBlock.CertifyingQC)
```
can we move this check so we execute it before `store`?

AlexHentschel replied:
👉 https://github.com/onflow/flow-go/pull/4190/commits/2231c4a0b0d3c2fec4a442b088b68e5142216ec8
(required some moderate restructuring of logic to avoid duplicating code or different behaviours of `AddProposal` vs `AddCertifiedBlock`)

jordanschalm commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+	parentBlock := parentVertex.(*BlockContainer2).Block()
+
+	// Note: we assume that all stored blocks pass Forks.EnsureBlockIsValidExtension(block);
+	//       specifically, that Proposal's ViewNumber is strictly monotonously
```
```suggestion
	//       specifically, that Proposal's ViewNumber is strictly monotonically
```

AlexHentschel replied:
Fixed the incorrect usages within Forks 👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

jordanschalm commented on consensus/hotstuff/forks/forks2_test.go:
```diff
@@ -0,0 +1,797 @@
+// TestDoubleProposal tests that the DoubleProposal notification is emitted when two different
+// proposals for the same view are added. We ingest the the following block tree:
+//
+//		    / [(◄1) 2]
+//		[1]
+//	       \ [(◄1) 2']
```
```suggestion
//                / [(◄1) 2]
//		[1]
//                \ [(◄1) 2']
```

jordanschalm reviewed (approved):
Nice work. No major change requests: many documentation suggestions and a few small functional suggestions.

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+	if err != nil {
+		return fmt.Errorf("emitting finalization events up to block %v failed: %w", qcForParent.BlockID, err)
+	}
+	f.finalityProof = &FinalityProof{Block: parentBlock, CertifiedChild: *certifiedBlock}
```
What I am worried the most is reentrancy, basically even if we are dealing with single threaded context, there is no guarantee that event consumer won't try to access `Forks`, it's perfectly legal and in the same thread, but accesses not yet updated fields.

AlexHentschel replied:
My original preference was also to update the `finalityProof` first and then emit the notifications first. I agree with your arguments @durkmurder: the notifications are anyway "eventually consistent", so consumers expect those notifications to lag behind direct polls of the same information. While there could be potential problems with re-entrancy (even for single-threaded callers).

Unfortunately,  first emitting the notifications _significantly_ simplifies the algorithm, for [the reason Jordan explained](https://github.com/onflow/flow-go/pull/4190#discussion_r1163172142):
>  - `finalizationEventsUpToBlock` relies on `FinalizedBlock()` not being updated when it is called in its implementation

Let me see if I can solve this 🤔

AlexHentschel replied:
I think I found a solution: https://github.com/onflow/flow-go/pull/4190/commits/813a8188ae6d9c4f906955f03b6173ccc6d00480 🤞

durkmurder commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+// finalization rules as defined in Jolteon consensus https://arxiv.org/abs/2106.10362
+// The same approach has later been adopted by the Diem team resulting in DiemBFT v4:
+// https://developers.diem.com/papers/diem-consensus-state-machine-replication-in-the-diem-blockchain/2021-08-17.pdf
+// Forks is NOT safe for concurrent use by multiple goroutines.
```
My understanding is that we will still deal with `FollowerLoop`, but removing one layer of abstraction.
Currently we have:
`Compliance` -> `FollowerLoop` -> `FollowerLogic` -> `Forks`.
We can get rid of `FollowerLogic` and ingest directly in `Forks`. `FollowerLoop` still guarantees seq access.

AlexHentschel replied:
☝️ exactly what Yurii said ... at least that is my plan for now ... but that could still turn out to be very complicated (but I am not expecting it to be, already went over the code in detail before I started all this work)

AlexHentschel commented on consensus/hotstuff/model/errors.go:
```diff
@@ -170,6 +170,11 @@ type InvalidBlockError struct {
 
+// NewInvalidBlockError instantiates an `InvalidBlockError`. Input `err` cannot be nil.
+func NewInvalidBlockError(blockID flow.Identifier, view uint64, err error) error {
+	return &InvalidBlockError{BlockID: blockID, View: view, Err: err}
```
💡 thanks for noticing

AlexHentschel replied:
implemented in https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+		v := vertexIterator.NextVertex()
+		l = append(l, v.(*BlockContainer2).Block())
+	}
+	return l
```
implemented in https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+// of blocks already stored (no state modifications). Specifically, the following condition
+// are enforced, which are critical to the correctness of Forks:
+//
+//  1. If block with the same ID is already stored, their views must be identical.
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+//	      compatible (principle of vacuous truth), i.e. we skip checking 1, 2, 3.
+//	 (ii) If block.View == F, we do not inspect the QC / parent at all (skip 2 and 3).
+//	      This exception is important for compatability with genesis or spork-root blocks,
+//	      which not contain a QCs.
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+	if (block.View == f.forest.LowestLevel) || (block.QC.View < f.forest.LowestLevel) { // exclusion (ii) and (iii)
+		return nil
+	}
+	// for block whose parents are _not_ below the pruning height, we expect the parent to be known.
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+// AddCertifiedBlock appends the given certified block to the tree of pending
+// blocks and updates the latest finalized block (if finalization progressed).
+// Unless the parent is below the pruning threshold (latest finalized view), we
+// require that he parent is already stored in Forks.
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+}
+
+// store adds the given block to our internal `forest`, updates `newestView` (if applicable),
+// and emits an `OnBlockIncorporated` notifications. While repeated inputs yield result in
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+// In case a conflicting QC is found, an ByzantineThresholdExceededError is returned.
+//
+// Two Quorum Certificates q1 and q2 are defined as conflicting iff:
+//   - q1.View == q2.View
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

AlexHentschel commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -0,0 +1,557 @@
+func (f *Forks2) checkForDoubleProposal(block *model.Block) {
+	it := f.forest.GetVerticesAtLevel(block.View)
+	for it.HasNext() {
+		otherVertex := it.NextVertex() // by construction, must have same view as parentView
```
👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2

durkmurder reviewed (approved):
Looks really solid. I've left some comments about structuring and checks, if you disagree then I am happy to leave it as it is at least for now and we can get back to it later. Well done!

AlexHentschel commented:
I think I found a good solution to the [re-entrancy problem for Forks](https://github.com/onflow/flow-go/pull/4190#discussion_r1162787334). @jordanschalm @durkmurder if you have time to take another look that would be great. There are some notable revisions of Forks that accumulated addressing the review comments. Thanks

jordanschalm commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -197,147 +191,94 @@ func (f *Forks2) EnsureBlockIsValidExtension(block *model.Block) error {
+	err := f.checkForByzantineEvidence(certifiedBlock.Block)
 	if err != nil {
-		return fmt.Errorf("storing block %v in Forks failed %w", certifiedBlock.Block.BlockID, err)
+		return fmt.Errorf("cannot store certified block %v: %w", certifiedBlock.Block.BlockID, err)
```
```suggestion
		return fmt.Errorf("cannot check for Byzantine evidence in certified block %v: %w", certifiedBlock.Block.BlockID, err)
```

jordanschalm commented on consensus/hotstuff/forks/forks2.go:
```diff
@@ -197,147 +191,94 @@ func (f *Forks2) EnsureBlockIsValidExtension(block *model.Block) error {
+	err := f.checkForByzantineEvidence(proposal)
 	if err != nil {
-		return fmt.Errorf("storing block %v in Forks failed %w", block.BlockID, err)
+		return fmt.Errorf("cannot store block %v: %w", proposal.BlockID, err)
```
```suggestion
		return fmt.Errorf("cannot check Byzantine evidence for block %v: %w", proposal.BlockID, err)
```

AlexHentschel commented:
bors merge
